### ISSUE TITLE
Add a way to specify custom generators for columns

### ISSUE THREAD

**Issue by mhauru** [alan-turing-institute/sqlsynthgen#11]
By default sqlsynthgen generates data based on the data type of the column. This needs to be expanded to give the users the opportunity to affect the generation of particular columns (or sets of columns, for providing correlations), to increase fidelity. We (Markus & Iain) think that the minimum would be providing the option for writing custom generator functions for particular columns, to be able to say things like "this column of type text should have English first names" or "this column of type float should have values uniformly randomly from 110-150". More customisability for the generator class might also be necessary, though this is unclear.

We had a conversation on 2022-12-15 about whether to provide this option through
1) Generating .py files that would have the code for the generator classes, that the user can then go and edit. The advantage of this is the unlimited range of customisation for the user.
2) Create the generator classes programmatically with some metaprogramming, without writing them to a .py file, with the user giving a dictionary of `column_name: generator_function`. The advantage of this is avoiding dealing with strings of python code (have to think about indents and such) and dealing more gracefully with situations where a user has provided custom generators functions but wants to run sqlsynthgen again because e.g. the underlying schema has changed.

The conversation was inconclusive and needs a second round.

**Comment by mhauru** [comment]
From stand-up on 2022-12-16: For non-junk data, the first table to target for UCLH is demographics (ethnicity, age, etc.). Next up observation types and then observations. The observations table would benefit from input from real data, the others not that much. Also of interest: Locations (can be a copy of what real data has).

**Comment by myyong** [comment]
As an MVP, we have generic generators for random values ala mimesis. We would like to define and deploy custom generators. This is interlinked with our discussion whether we generate values by column or by table. This is recorded [here](https://github.com/alan-turing-institute/sqlsynthgen/issues/4).

We explored two options for enabling user selection/customisation of generators. 

1. One option involves the use of YAML files to define the generator function. This option has the challenge that we cannot generate related values eg (diastolic vs systolic values, visit start, end dates and duration). This option would be acceptable if we generated values by columns.

2. Another option involves writing custom generators for tables and using inheritance to override generic base generators. This option would be preferable if we generated values by tables. This option gives is the ability to use both base and custom generators.

**Comment by myyong** [comment]
We also discussed how to address the challenge of vocabulary tables. Base on [this](https://github.com/alan-turing-institute/sqlsynthgen/issues/11#issuecomment-1354484599) I am more convinced that hydef values provide the vocabulary for EMAP. I will check with Jon.

**Comment by mhauru** [comment]
We chose option 2., of the ones above.

There was also a question of whether we should change our class structure so that rather than creating a new generator instance for each row, there would be one persistent one for the whole table. This would allow having a memory for the table, so that we could for instance enforce some timestamp column to grow monotonously. We decided to postpone this decision, and stick with the current one-instance-per-row model for now, just to take one step at a time.
